spec/models/dynamic_field_group_spec.rb

```diff
@@ -0,0 +1,339 @@
+          parent.reload
+        end
+
+        it 'does not save' do
+          expect(dynamic_field_group.save).to be false
+        end
+
+        it 'returns correct error' do
+          dynamic_field_group.save
+          expect(dynamic_field_group.errors.full_messages).to include 'String key has already been taken'
+        end
+      end
+    end
+
+    context 'when creating a group that shares the same string_key as a non-sibling group' do
+      let(:dynamic_field_group) { FactoryBot.create(:dynamic_field_group) }
+      let(:dynamic_field) { FactoryBot.build(:dynamic_field, string_key: 'name', dynamic_field_group: dynamic_field_group) }
+
+      it 'saves the object' do
+        expect(dynamic_field_group.save).to be true
+      end
+    end
+  end
+
+  describe '#parent' do
+    let(:dynamic_field_group) { FactoryBot.build(:dynamic_field_group, :child) }
+
+    before do
+      dynamic_field_group.parent = parent
+      dynamic_field_group.save
+      parent.reload
+    end
+
+    context 'when setting DynamicFieldCategory as parent' do
+      let(:parent) { FactoryBot.create(:dynamic_field_category) }
+
+      it 'adds parent' do
+        expect(dynamic_field_group.parent).to eq parent
+      end
+
+      it 'adds dynamic field group to dynamic field category' do
+        expect(parent.dynamic_field_groups).to match_array [dynamic_field_group]
+      end
+    end
+
+    context 'when setting DynamicFieldGroup as parent' do
+      let(:parent) { FactoryBot.create(:dynamic_field_group) }
+
+      it 'adds parent' do
+        expect(dynamic_field_group.parent).to eq parent
+      end
+
+      it 'adds dynamic field group to parent dynamic field group' do
+        expect(parent.dynamic_field_groups).to match_array [dynamic_field_group]
+      end
+    end
+
+    context 'when setting parent to invalid parent type' do
+      let(:parent) { FactoryBot.create(:user, email: 'random_user@example.com') }
+
+      it 'does save' do
+        expect(dynamic_field_group.save).to be false
+      end
+
+      it 'returns correct error' do
+        dynamic_field_group.save
+        expect(dynamic_field_group.errors.full_messages).to include 'Parent type is not included in the list'
+      end
+    end
+  end
+
+  describe '#dynamic_fields' do
+    let(:dynamic_field_group) { FactoryBot.create(:dynamic_field_group) }
+    let(:dynamic_field_1) { FactoryBot.build(:dynamic_field, dynamic_field_group: nil) }
+    let(:dynamic_field_2) { FactoryBot.build(:dynamic_field, dynamic_field_group: nil, string_key: 'usage_primary') }
+
+    it 'can add a dynamic field' do
+      dynamic_field_group.dynamic_fields << dynamic_field_1
+      dynamic_field_1.save
+      expect(dynamic_field_group.dynamic_fields).to match_array [dynamic_field_1]
+    end
+
+    it 'can add multiple dynamic fields' do
+      dynamic_field_group.dynamic_fields << dynamic_field_1
+      dynamic_field_group.dynamic_fields << dynamic_field_2
+      dynamic_field_1.save
```

Comment:
Same comment here as above, about `dynamic_field_1.save` not adding anything to this test.  I think the test would still function the same way without this `.save` call and the one below it.

Reply:
I don't think they would because I just build the factories, not create them. I just wanted to test the addition of new fields. I am adding, saving, reloading and then test to make sure they are present.

Reply:
I'll change save to save!, per our discussion